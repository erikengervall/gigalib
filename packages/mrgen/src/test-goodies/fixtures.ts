import { MRGenOptions } from '../types/mrgen-options';

export const createMockMRGenOptions = (
  overrides?: Partial<MRGenOptions>,
): { mrGenOptions: MRGenOptions } => {
  return {
    mrGenOptions: {
      data: {
        queryParams: true,
        payload: true,
        params: true,
        responseBody: true,
      },

      httpMethod: 'POST',

      shared: {
        apiDir: 'packages/shared/src/api',
        file: 'packages/shared/src/api/v1/user/get.ts',
        buildCommand: 'yarn workspace shared ci:build',
      },
      backend: {
        file: 'packages/backend/src/handlers/v1/user/get-handler.ts',
      },
      frontend: {
        file: 'packages/frontend1/src/api/hooks/v1-user-get.ts',
      },
      ...overrides,
    },
  };
};
