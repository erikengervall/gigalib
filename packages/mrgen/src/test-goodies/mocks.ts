import cp from 'node:child_process';
import fs from 'node:fs';

export const mockExecSync = cp.execSync as jest.Mock;
export const mockMkdirSync = fs.mkdirSync as jest.Mock;
export const mockWriteFileSync = fs.writeFileSync as jest.Mock;
