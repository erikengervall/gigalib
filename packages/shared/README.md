# @engervall/shared

Utilities and stuff.
