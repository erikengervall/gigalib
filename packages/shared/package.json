{
  "name": "@engervall/shared",
  "version": "1.0.6-alpha.0",
  "main": "dist/index.js",
  "repository": "git@github.com:erikengervall/gigalib.git",
  "author": "Erik Engervall <erik.engervall@gmail.com>",
  "scripts": {
    "----- CI": "",
    "ci:build": "tsc",
    "ci:format": "prettier --check './src/**/*.{ts,tsx}' --ignore-path ../../.prettierignore",
    "ci:lint": "eslint 'src/**/*.{ts,tsx}'",
    "ci:test:unit": "jest",
    "----- LOCAL DEV": "",
    "dev": "rm -rf dist && tsc --watch",
    "----- MISC": "",
    "format:write": "prettier --write './src/**/*.{ts,tsx}'",
    "clean": "rm -rf dist",
    "reinstall": "yarn clean && rm -rf node_modules"
  },
  "dependencies": {
    "qs": "^6.11.2",
    "zod": "^3.21.4",
    "zod-validation-error": "^1.3.1"
  }
}
