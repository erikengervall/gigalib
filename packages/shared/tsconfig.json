{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "lib": ["es2021"],
    "module": "CommonJS",
    "noEmit": false,
    "sourceMap": true,
    "target": "ES6",

    "baseUrl": "src",
    "outDir": "dist",
    "rootDir": "src"
  },
  "include": ["src", "../../scripts/change-case.ts"],
  "exclude": [
    "_warmup",
    ".cache",
    ".next",
    ".serverless",
    ".vscode",
    ".vscode",
    ".webpack",
    "*.dist*",
    "*.spec*",
    "*.test*",
    "build",
    "coverage",
    "cypress-coverage",
    "dist",
    "jest-coverage",
    "node_modules"
  ]
}
