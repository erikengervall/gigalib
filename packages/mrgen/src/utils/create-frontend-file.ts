import { pascalCase } from '@engervall/shared';

import { MRGenOptions } from '../types/mrgen-options';
import { createFileWithContent } from './create-file-with-content';
import { formatFile } from './format-file';
import { getAPIType } from './get-api-type';

export function createFrontendFile(options: MRGenOptions) {
  const { APIType } = getAPIType(options);

  const hookName = pascalCase(APIType.base);
  // const queryVarName = `use${hookName}Query`;

  let frontendFileContent = '';
  const responseBody = options.data.responseBody
    ? APIType.responseBody
    : 'void';

  switch (options.httpMethod) {
    case 'POST': {
      const mutationVarName = `use${hookName}Mutation`;
      const mutationInput = (
        options.data.payload || options.data.params || options.data.queryParams
          ? `{
              ${options.data.payload ? `payload: ${APIType.payload};` : ''}
              ${options.data.params ? `params: ${APIType.params};` : ''}
              ${options.data.queryParams ? `queryParams: ${APIType.queryParams};` : ''}
            }`
          : 'void'
      )
        .split('\n')
        .join('');

      frontendFileContent = `
    import { apiClient } from './api-client';
    import type { FrontendResponseError } from '@engervall/shared';
    import { UseMutationOptions, useMutation } from '@tanstack/react-query';
    import { API } from 'shared';
    ${options.data.queryParams ? "import QueryString from 'qs';" : ''}
    
    export const ${mutationVarName} = (
      options?: UseMutationOptions<
        ${responseBody},
        FrontendResponseError,
        ${mutationInput}
      >,
    ) => {
      return useMutation({
        mutationFn: async (input) => {
          ${options.data.queryParams ? 'const queryParams = QueryString.stringify(input.queryParams);' : ''}
          const response = await apiClient.post<${responseBody}>(
            \`${APIType.urlPath}${options.data.queryParams ? '?${queryParams}' : ''}\`,
            ${options.data.payload ? 'input.payload' : 'undefined'},
            ${options.data.params ? '{ params: input.params }' : ''}
          );
          return response.data;
        },
        mutationKey: ['${mutationVarName}'],
        ...options,
      });
    };
    `;
      break;
    }

    case 'GET': {
      const queryVarName = `use${hookName}Query`;
      const queryInput = (
        options.data.params || options.data.queryParams
          ? `{
              ${options.data.params ? `params: ${APIType.params};` : ''}
              ${options.data.queryParams ? `queryParams: ${APIType.queryParams};` : ''}
            }`
          : 'void'
      )
        .split('\n')
        .join('');

      frontendFileContent = `
    import { apiClient } from './api-client';
    import type { FrontendResponseError } from '@engervall/shared';
    import { UseQueryOptions, useQuery } from '@tanstack/react-query';
    import { API } from 'shared';
    ${options.data.queryParams ? "import QueryString from 'qs';" : ''}

    export const ${queryVarName} = (
      input: ${queryInput},
      options?: UseQueryOptions<
        ${responseBody},
        FrontendResponseError
      >,
    ) => {
      return useQuery({
        queryFn: async () => {
          ${options.data.queryParams ? 'const queryParams = QueryString.stringify(input.queryParams);' : ''}
          const response = await apiClient.get<${responseBody}>(
            \`${APIType.urlPath}${options.data.queryParams ? '?${queryParams}' : ''}\`,
            ${options.data.params ? '{ params: input.params }' : ''}
          );
          return response.data;
        },
        queryKey: ['${queryVarName}'],
        ...options,
      });
    };
`;
      break;
    }

    default: {
      throw new Error(`Unsupported httpMethod: ${options.httpMethod}`);
    }
  }

  createFileWithContent(options.frontend.file, frontendFileContent);

  formatFile(options.frontend.file);
}
