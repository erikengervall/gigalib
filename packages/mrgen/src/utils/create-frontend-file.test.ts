import { logSnapshots } from '@engervall/shared';

import { createMockMRGenOptions } from '../test-goodies/fixtures';
import {
  mockExecSync,
  mockMkdirSync,
  mockWriteFileSync,
} from '../test-goodies/mocks';
import { createFrontendFile } from './create-frontend-file';

it('should create a frontend file for a GET request', () => {
  const { mrGenOptions } = createMockMRGenOptions({
    httpMethod: 'GET',
  });

  createFrontendFile(mrGenOptions);

  expect(mockMkdirSync.mock.calls).toMatchSnapshot('fs.mkdirSync');
  expect(mockWriteFileSync.mock.calls).toMatchSnapshot('fs.writeFileSync');
  expect(JSON.stringify(mockWriteFileSync.mock.calls)).toContain('useQuery');
  expect(mockExecSync.mock.calls).toMatchSnapshot('cp.execSync');
  logSnapshots();
});

it('should create a frontend file for a POST request', () => {
  const { mrGenOptions } = createMockMRGenOptions({
    httpMethod: 'POST',
  });

  createFrontendFile(mrGenOptions);

  expect(mockMkdirSync.mock.calls).toMatchSnapshot('fs.mkdirSync');
  expect(mockWriteFileSync.mock.calls).toMatchSnapshot('fs.writeFileSync');
  expect(JSON.stringify(mockWriteFileSync.mock.calls)).toContain('useMutation');
  expect(mockExecSync.mock.calls).toMatchSnapshot('cp.execSync');
  logSnapshots();
});
