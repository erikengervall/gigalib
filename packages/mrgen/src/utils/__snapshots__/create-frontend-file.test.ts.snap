// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a frontend file for a GET request: aggregateLogProxySpy 1`] = `
[
  {
    "debug": [
      "mkdirSync",
      "packages/frontend1/src/api/hooks",
    ],
  },
  {
    "debug": [
      "writeFileSync",
      "packages/frontend1/src/api/hooks/v1-user-get.ts",
    ],
  },
  {
    "debug": [
      "Running command",
      "yarn prettier --write packages/frontend1/src/api/hooks/v1-user-get.ts",
    ],
  },
]
`;

exports[`should create a frontend file for a GET request: cp.execSync 1`] = `
[
  [
    "yarn prettier --write packages/frontend1/src/api/hooks/v1-user-get.ts",
  ],
]
`;

exports[`should create a frontend file for a GET request: fs.mkdirSync 1`] = `
[
  [
    "packages/frontend1/src/api/hooks",
    {
      "recursive": true,
    },
  ],
]
`;

exports[`should create a frontend file for a GET request: fs.writeFileSync 1`] = `
[
  [
    "packages/frontend1/src/api/hooks/v1-user-get.ts",
    "
    export const useApiV1UserGetQuery = (
      input: {              params: API.V1.User.Get.Params;              queryParams: API.V1.User.Get.QueryParams;            },
      options?: UseQueryOptions<
        API.V1.User.Get.ResponseBody,
        FrontendResponseError
      >,
    ) => {
      return useQuery({
        queryFn: async () => {
          const queryParams = QueryString.stringify(input.queryParams);
          const response = await apiClient.get<API.V1.User.Get.ResponseBody>(
            \`api/v1/user/get?\${queryParams}\`,
            { params: input.params }
          );
          return response.data;
        },
        queryKey: ['useApiV1UserGetQuery'],
        ...options,
      });
    };
",
    "utf-8",
  ],
]
`;

exports[`should create a frontend file for a POST request: aggregateLogProxySpy 1`] = `
[
  {
    "debug": [
      "mkdirSync",
      "packages/frontend1/src/api/hooks",
    ],
  },
  {
    "debug": [
      "writeFileSync",
      "packages/frontend1/src/api/hooks/v1-user-get.ts",
    ],
  },
  {
    "debug": [
      "Running command",
      "yarn prettier --write packages/frontend1/src/api/hooks/v1-user-get.ts",
    ],
  },
]
`;

exports[`should create a frontend file for a POST request: cp.execSync 1`] = `
[
  [
    "yarn prettier --write packages/frontend1/src/api/hooks/v1-user-get.ts",
  ],
]
`;

exports[`should create a frontend file for a POST request: fs.mkdirSync 1`] = `
[
  [
    "packages/frontend1/src/api/hooks",
    {
      "recursive": true,
    },
  ],
]
`;

exports[`should create a frontend file for a POST request: fs.writeFileSync 1`] = `
[
  [
    "packages/frontend1/src/api/hooks/v1-user-get.ts",
    "
    import { apiClient } from '../api-client';
    import type { FrontendResponseError } from '../frontend-response-error';
    import { UseMutationOptions, useMutation } from '@tanstack/react-query';
    import { API } from 'shared';
    import QueryString from 'qs';
    
    export const useApiV1UserGetMutation = (
      options?: UseMutationOptions<
        API.V1.User.Get.ResponseBody,
        FrontendResponseError,
        {              payload: API.V1.User.Get.Payload;              params: API.V1.User.Get.Params;              queryParams: API.V1.User.Get.QueryParams;            }
      >,
    ) => {
      return useMutation({
        mutationFn: async (input) => {
          const queryParams = QueryString.stringify(input.queryParams);
          const response = await apiClient.post<API.V1.User.Get.ResponseBody>(
            \`api/v1/user/get?\${queryParams}\`,
            input.payload,
            { params: input.params }
          );
          return response.data;
        },
        mutationKey: ['useApiV1UserGetMutation'],
        ...options,
      });
    };
    ",
    "utf-8",
  ],
]
`;
